cli/src/commands/rebase.rs

```diff
@@ -348,91 +351,107 @@ fn rebase_descendants_transaction(
+            return Err(user_error(format!(
+                "Cannot rebase {} onto itself",
+                short_commit_hash(commit.id()),
+            )));
+        }
     }
 
     let mut tx = workspace_command.start_transaction();
-    let mut rebased_commit_ids = HashMap::new();
+    let tx_description = if target_commits.len() == 1 {
+        format!("rebase commit {}", target_commits[0].id().hex())
+    } else {
+        format!(
+            "rebase commit {} and {} more",
+            target_commits[0].id().hex(),
+            target_commits.len() - 1
+        )
+    };
+
+    // First, rebase the descendants of `target_commits`.
+    let (initial_rebased_commit_ids, target_roots) =
+        extract_commits(&mut tx, settings, &target_commits)?;
+
+    // We now update `new_parents` to account for the rebase of all of
+    // `target_commits`'s descendants. Even if some of the original `new_parents`
+    // were descendants of `target_commits`, this will no longer be the case after
+    // the update.
+    let new_parents = new_parents
+        .iter()
+        .map(|new_parent| {
+            get_possibly_rewritten_commit_id(&initial_rebased_commit_ids, new_parent.id())
+        })
+        .collect_vec();
+    let mut new_rebased_commit_ids = HashMap::new();
+    let mut skipped_commits = Vec::new();
 
-    // First, rebase the descendants of `to_rebase_commit`.
-    // TODO(ilyagr): Consider making it possible for these descendants to become
-    // emptied, like --skip_empty. This would require writing careful tests.
+    // At this point, all commits in the target set will only have other commits in
+    // the set as their ancestors. We can now safely rebase `target_commits` onto
+    // the `new_parents`, by updating the roots' parents and rebasing its
+    // descendants.
     tx.mut_repo().transform_descendants(
         settings,
-        vec![to_rebase_commit.id().clone()],
+        target_roots.iter().cloned().collect_vec(),
         |mut rewriter| {
             let old_commit = rewriter.old_commit();
             let old_commit_id = old_commit.id().clone();
 
-            // Replace references to `to_rebase_commit` with its parents.
-            rewriter.replace_parent(to_rebase_commit.id(), to_rebase_commit.parent_ids());
+            if target_roots.contains(&old_commit_id) {
+                rewriter.set_new_parents(new_parents.clone());
+            }
             if rewriter.parents_changed() {
                 let builder = rewriter.rebase(settings)?;
                 let commit = builder.write()?;
-                rebased_commit_ids.insert(old_commit_id, commit.id().clone());
+                new_rebased_commit_ids.insert(old_commit_id, commit.id().clone());
+            } else {
+                // Only include commit in skipped commits if it wasn't previously rewritten.
+                if target_commits
+                    .iter()
+                    .any(|commit| commit.id() == rewriter.old_commit().id())
```

Comment:
nit: build `target_commit_ids: HashSet<CommitId>` upfront, and pass it to `extract_commits()`?

Reply:
Updated.